Remove hooks from ERC20 (#3838)
OpenZeppelin/openzeppelin-contracts: CHANGELOG.md

 * Removed presets in favor of [OpenZeppelin Contracts Wizard](https://wizard.openzeppelin.com/).
 * `TransparentUpgradeableProxy`: Removed `admin` and `implementation` getters, which were only callable by the proxy owner and thus not very useful. ([#3820](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3820))
 * `ProxyAdmin`: Removed `getProxyAdmin` and `getProxyImplementation` getters. ([#3820](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3820))
 * `ERC20`: Deleted `_beforeTokenTransfer` and `_afterTokenTransfer` hooks, and refactor all extensions using those hooks for customization, by using `_update` instead. ([#3838](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3838))

### How to upgrade from 4.x

#### ERC20, ERC721, and ERC1155

> These breaking changes will require modifications to ERC20, ERC721, and ERC1155 contracts, since the `_afterTokenTransfer` and `_beforeTokenTransfer` functions were removed. Any customization made through those hooks should now be done overriding the `_transfer` function instead.
> 
> Minting and burning are implemented by `_transfer` and customizations should be done by overriding this function as well. `_mint` and `_burn` are no longer virtual (meaning they are not overridable) to guard against possible inconsistencies.

Reviewer: ```suggestion
These breaking changes will require modifications to ERC20, ERC721, and ERC1155 contracts, since the `_afterTokenTransfer` and `_beforeTokenTransfer` functions were removed. Any customization made through those hooks should now be done overriding the new `_update` function instead.

Minting and burning are implemented by `_update` and customizations should be done by overriding this function as well. `_mint` and `_burn` are no longer virtual (meaning they are not overridable) to guard against possible inconsistencies.
```

For example, a contract using `ERC20`'s `_afterTokenTransfer` hook would have to be changed in the following way.

```diff
- function _beforeTokenTransfer(

> + function _transfer(
>       address from,
>       address to,
>       uint256 amount
>   ) internal virtual override {
> -     super._beforeTokenTransfer(from, to, amount);
>       require(!condition(), "ERC20: wrong condition");
> +     super._transfer(from, to, amount);

Reviewer: ```suggestion
+ function _update(
      address from,
      address to,
      uint256 amount
  ) internal virtual override {
-     super._beforeTokenTransfer(from, to, amount);
      require(!condition(), "ERC20: wrong condition");
+     super._update(from, to, amount);
```

 * `ProxyAdmin`: Removed `getProxyAdmin` and `getProxyImplementation` getters. ([#3820](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3820))
 * `ERC20`: Deleted `_beforeTokenTransfer` and `_afterTokenTransfer` hooks, and refactor all extensions using those hooks for customization, by using `transfer` instead. ([#3838](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3838))

### How to upgrade from 4.x
These breaking changes will require additional modifications in your implementation of ERC20, ERC721, and ERC1155, since the `afterTokenTransfer` and `beforeTokenTransfer` were removed. Any customization made through those hooks should now use the `transfer` function instead, keeping in mind that `_mint` and `_burn` are no longer virtual, so they are not overridable.

> For example, a contract using `ERC20`'s `afterTokenTransfer` hook would have to be changed in the following way.

Reviewer: ```suggestion
For example, a contract using `ERC20`'s `_afterTokenTransfer` hook would have to be changed in the following way.
```

```diff

> -function _beforeTokenTransfer(
> -         address from,
> -         address to,
> -         uint256 amount
> -     ) internal virtual override {
> -         super._beforeTokenTransfer(from, to, amount);
> -
> -         require(!condition(), "ERC20: wrong condition");
> -     }
> +function _transfer(
> +         address from,
> +         address to,
> +         uint256 amount
> +     ) internal virtual override {
> +         require(!condition(), "ERC20: wrong condition");
> +         super._transfer(from, to, amount);
> +     }

Reviewer: This should be simpler and clearer:

```diff
- function _beforeTokenTransfer(
+ function _transfer(
      address from,
      address to,
      uint256 amount
  ) internal virtual override {
-     super._beforeTokenTransfer(from, to, amount);
      require(!condition(), "ERC20: wrong condition");
+     super._transfer(from, to, amount);
  }
```

The below suggestion implements this

```suggestion
- function _beforeTokenTransfer(
+ function _transfer(
      address from,
      address to,
      uint256 amount
  ) internal virtual override {
-     super._beforeTokenTransfer(from, to, amount);
      require(!condition(), "ERC20: wrong condition");
+     super._transfer(from, to, amount);
  }
```
